Source/Scene/ShadowMapShader.js

```diff
@@ -0,0 +1,368 @@
+
+        var usesDepthTexture = shadowMap._usesDepthTexture;
+        var isPointLight = shadowMap._isPointLight;
+        var isSpotLight = shadowMap._isSpotLight;
+        var usesCubeMap = shadowMap._usesCubeMap;
+        var hasCascades = shadowMap._numberOfCascades > 1;
+        var cascadesEnabled = shadowMap._cascadesEnabled;
+        var debugVisualizeCascades = shadowMap.debugVisualizeCascades;
+        var softShadows = shadowMap.softShadows;
+        var bias = isPointLight ? shadowMap._pointBias : (isTerrain ? shadowMap._terrainBias : shadowMap._primitiveBias);
+        var exponentialShadows = shadowMap._exponentialShadows;
+
+        var defines = fs.defines.slice(0);
+        var sources = fs.sources.slice(0);
+
+        var length = sources.length;
+        for (var i = 0; i < length; ++i) {
+            sources[i] = ShaderSource.replaceMain(sources[i], 'czm_shadow_receive_main');
+        }
+
+        if (isPointLight && usesCubeMap) {
+            defines.push('USE_CUBE_MAP_SHADOW');
+        } else if (usesDepthTexture) {
+            defines.push('USE_SHADOW_DEPTH_TEXTURE');
+        }
+
+        if (softShadows && !isPointLight) {
+            defines.push('USE_SOFT_SHADOWS');
+        }
+
+        if (castShadows && bias.normalShading && hasNormalVarying) {
+            defines.push('USE_NORMAL_SHADING');
+            if (bias.normalShadingSmooth > 0.0) {
+                defines.push('USE_NORMAL_SHADING_SMOOTH');
+            }
+        }
+
+        if (exponentialShadows) {
+            defines.push('USE_EXPONENTIAL_SHADOW_MAPS');
+        }
+
+        var fsSource =
+            'uniform mat4 u_shadowMapMatrix; \n' +
+            'uniform vec3 u_shadowMapLightDirectionEC; \n' +
+            'uniform vec4 u_shadowMapLightPositionEC; \n' +
+            'uniform vec4 u_shadowMapNormalOffsetScaleDistanceMaxDistanceAndDarkness; \n' +
+            'uniform vec4 u_shadowMapTexelSizeDepthBiasAndNormalShadingSmooth; \n' +
+            'uniform sampler2D u_shadowMapTexture; \n' +
+            'uniform samplerCube u_shadowMapTextureCube; \n' +
+            'vec4 getPositionEC() \n' +
+            '{ \n' +
+            (hasPositionVarying ?
+            '    return vec4(' + positionVaryingName + ', 1.0); \n' :
+            '    return czm_windowToEyeCoordinates(gl_FragCoord); \n') +
+            '} \n' +
+            'vec3 getNormalEC() \n' +
+            '{ \n' +
+            (hasNormalVarying ?
+            '    return normalize(' + normalVaryingName + '); \n' :
+            '    return vec3(1.0); \n') +
+            '} \n' +
+
+            'void applyNormalOffset(inout vec4 positionEC, vec3 normalEC, float nDotL) \n' +
+            '{ \n' +
+            (bias.normalOffset && hasNormalVarying ?
+            '    float normalOffset = u_shadowMapNormalOffsetScaleDistanceMaxDistanceAndDarkness.x; \n' +
+            '    // Offset the shadow position in the direction of the normal for perpendicular and back faces \n' +
```

Comment:
Do you think it is better to make this a JavaScript comment so it gets syntax highlighting?